net, ctrl, vm: Consolidate hotplug / hotunplug logic

orelmisan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
This is a follow-up to https://github.com/kubevirt/kubevirt/pull/13901 and https://github.com/kubevirt/kubevirt/pull/13844.

Currently, there are two unit test files testing the NIC hotplug / hotunplug logic:
- pkg/network/controllers/vm_test.go
- pkg/network/controllers/hotplug_test.go

Move tests testing lower level logic from `hotplug_test.go` to the higher level `vm_test.go` file that tests the network VM controller as a whole.

Move `ApplyDynamicIfaceRequestOnVMI` and `ClearDetachedInterfaces` closer to their only caller (network VM controller).

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

orelmisan commented:
/kind cleanup

orelmisan commented:
/cc @nirdothan

EdDev reviewed (commented):
>  net, ctrl: Mv ClearDetachedInterfaces
> 
> This function is only used by the network VM controller.
> Move it closer to its caller.
> 
> Its unit tests are covered by the network VM controller's unit tests[1].
> 
> [1] https://github.com/kubevirt/kubevirt/blob/c6aeb68d1794b71026236f726c15aa9550cc6d64/pkg/network/controllers/vm_test.go#L111

Please clarify which test is covering this scenario:
> "should remain, given status iface and no associated iface in spec"

RamLavi commented on pkg/network/controllers/vm_test.go:
```diff
@@ -509,6 +509,70 @@ var _ = Describe("VM Network Controller", func() {
+			libvmi.WithInterface(ifaceToDetach),
+			libvmi.WithNetwork(v1.DefaultPodNetwork()),
+			libvmi.WithNetwork(netToDetach),
+			libvmistatus.WithStatus(
```
this was not in the original test, can you explain why it is needed?

orelmisan replied:
The original test only tested a small fraction of the logic.
The proposed test, tests the whole VM controller's logic in the scenario when one interface is hotplugged and another is hotunplugged.
In the test, we hotunplug a device that is currently plugged (has an interface status with certain info sources),

RamLavi reviewed (commented):
Looks good overall, added some comments

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/d036d6c69b167683e7b43c5a9a06671b33ac19af..4f7d3b3a25be793887455e2274ba97cda1d731bb): Rebase.

orelmisan commented on pkg/network/controllers/vm_test.go:
```diff
@@ -156,6 +156,46 @@ var _ = Describe("VM Network Controller", func() {
 	)
 
+	It("sync does not hotplug a new absent interface", func() {
+		clientset := fake.NewSimpleClientset()
```
Agreed.

orelmisan replied:
I prefer to perform additional refactoring in follow-up PRs.
This PR has already became quite bit.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/4f7d3b3a25be793887455e2274ba97cda1d731bb..ad5e0b2f57a1fb5d4e29a04de1a4bae344e01fcf): Updated commit message in order to address @EdDev's [comment](https://github.com/kubevirt/kubevirt/pull/14419#pullrequestreview-2745060489).

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/ad5e0b2f57a1fb5d4e29a04de1a4bae344e01fcf..602f21c7c7483b4818e2745d2ff42f86232a52de): Fixed a typo in commit message.

orelmisan commented:
> > The "should remain, given status iface and no associated iface in spec"
> > case's title is misleading, it actually tests that a spec interface will
> > not be cleared in case there is no matching interface status.
> > This scenario is covered by the hotplug scenario [1] as well.
> 
> I do not see what is misleading about it. Given there is an interface reported in the status but not in the spec, it should do nothing. The test describes it correctly, the implementation used it mixed with another scenario which may caused this confusion.
> 
> Still, the coverage of this case needs to be shown or explained why not.

The test's full name is `spec interfaces should remain, given status iface and no associated iface in spec`.
It is given an interface with a matching network, and an unrelated interface status.
The test asserts that the spec interface and its network are still there following the execution of the production logic.
The test does not assert about the interface status, so I think that the hotplug test covers it since it has no matching interface status for the hotplugged spec interface and network.
Also, the newly added hotplug and hotunplug scenario covers this case since there it has a hotplugged interface without a matching interface status and there are additional unrelated interface statuses.

EdDev commented:
> and an unrelated interface status

The test name interprets it in the other direction: There is an interface on the status but not on the spec. For the interface that is specified on the spec, you can add the missing status and it won’t change the test (or what it aims to check).

ormergi commented on pkg/network/controllers/vm_test.go:
```diff
@@ -156,6 +156,46 @@ var _ = Describe("VM Network Controller", func() {
+
+		vm = plugNetworkInterface(vm, absentIfaceToPlug)
+
+		// Simulate the existence of the VMI on the server (to allow the Sync to patch it).
```
Using By reflect in the test logs allowing to troubleshoot when there are failures.
For example having `By("Assert that the hotplugged interface and its matching network were cleared")`
result in Gingko to print the following line 
```
STEP: Assert that the hotplugged interface and its matching network were cleared
```

In e2e this is very useful.

orelmisan replied:
> In e2e this is very useful.

I agree.
But keep in mind that these are unit tests, which quickly run to completion.
Having them around is not much different from having comments - which can rot and are sometimes a maintenance burden.

ormergi commented on pkg/network/controllers/vm_test.go:
```diff
@@ -509,6 +509,70 @@ var _ = Describe("VM Network Controller", func() {
+		vm = plugNetworkInterface(vm, libvmi.InterfaceDeviceWithBridgeBinding(netToAttachName))
+
+		// Simulate the existence of the VMI on the server (to allow the Sync to patch it).
+		_, err := clientset.KubevirtV1().VirtualMachineInstances(vmi.Namespace).Create(context.Background(), vmi, k8smetav1.CreateOptions{})
```
I think the test missing the VirtualMachine creation as well, having the test simulate VM existence should prevent unwanted side effects.

orelmisan replied:
The VM objects are not "created" on the fake client all over the file.
This is because the network VM controller does not directly fetches / mutates the VirtualMachine object (currently it just patches the VMI).

ormergi commented on pkg/network/controllers/vm_test.go:
```diff
@@ -509,6 +509,70 @@ var _ = Describe("VM Network Controller", func() {
+		updatedVM, err := c.Sync(vm, vmi)
+		Expect(err).NotTo(HaveOccurred())
+
+		Expect(updatedVM).To(Equal(originalVM))
```
nit: I think adding a comment saying "the controller should not change the VM object" would be nice.

orelmisan replied:
> nit: I think adding a comment saying "the controller should not change the VM object" would be nice.

Isn't it reflected with this assertion?

ormergi commented on pkg/network/controllers/hotplug_test.go:
```diff
@@ -1,68 +0,0 @@
-			Expect(ifaces).To(Equal(expectedInterfaces))
-			Expect(networks).To(Equal(expectedNetworks))
-		},
-		Entry("should remain, given non-absent interfaces, and no associated status ifaces (i.e.: plug pending)",
-			[]v1.Interface{{Name: "blue"}, {Name: "red"}},
-			[]v1.VirtualMachineInstanceNetworkInterface{},
-			[]v1.Interface{{Name: "blue"}, {Name: "red"}},
-			[]v1.Network{{Name: "blue"}, {Name: "red"}},
-		),
-		Entry("should remain, given non-absent interfaces, and associated status ifaces (i.e.: plugged iface)",
-			[]v1.Interface{{Name: "blue"}, {Name: "red"}},
-			[]v1.VirtualMachineInstanceNetworkInterface{{Name: "blue"}, {Name: "red"}},
-			[]v1.Interface{{Name: "blue"}, {Name: "red"}},
-			[]v1.Network{{Name: "blue"}, {Name: "red"}},
-		),
-		Entry("should remain, given status iface and no associated iface in spec",
-			[]v1.Interface{{Name: "blue"}},
-			[]v1.VirtualMachineInstanceNetworkInterface{{Name: "RED"}},
-			[]v1.Interface{{Name: "blue"}},
-			[]v1.Network{{Name: "blue"}},
-		),
```
The mentioned test [1] covers a successful scenario of hotplug, but it doesnt seem to cover these scenarios, when there are pending hotplug/unplug  and when a removed iface remain in status.
Please consider extending the [1] to add new one(s) to cover that.

[1] https://github.com/kubevirt/kubevirt/blob/c6aeb68d1794b71026236f726c15aa9550cc6d64/pkg/network/controllers/vm_test.go#L111

orelmisan replied:
The hotplug test has two interfaces without a matching interface status - Same as "should remain, given non-absent interfaces, and no associated status ifaces (i.e.: plug pending)".

The hotunplug test has a pending interface.

ormergi reviewed (commented):
Hi, thanks for this clean up, I like the idea hotplug/unplug is being tested by a higher abstraction layer.

One comment about hot the new test work, it seems the test calls the network controller Sync "statically", instead of having the controller actually running as we would with a real env.
This approach may result in some side-effects not being detected by the tests, vs. having a the controller actually running as it would in a living env.
Although I do understand changing the tests to work with a running controller require additional work. Pleas consider that for following PRs.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/602f21c7c7483b4818e2745d2ff42f86232a52de..3a133bc7c6955e4b52cb00287aba597ea397c84f): Rebase.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/3a133bc7c6955e4b52cb00287aba597ea397c84f..8a5786bde6b0e144e1a5ca693e4d1d0651ca87a9): Addressed review comments.

EdDev reviewed (commented):
Thanks!

/approve

orelmisan commented on pkg/network/controllers/vm_test.go:
```diff
@@ -156,6 +156,46 @@ var _ = Describe("VM Network Controller", func() {
+		clientset := fake.NewSimpleClientset()
+		c := controllers.NewVMController(clientset)
+		vmi := libvmi.New(
+			libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
```
> is it just me or we didnt have it on the previous test? if it is not a must maybe worth to drop ?

I don't follow, could you please elaborate?

oshoval replied:
previous test didn't have the Masquerade interface, right ?

orelmisan replied:
> previous test didn't have the Masquerade interface, right ? if so, might worth to drop it to have the test the minimal required

What is the name of the test in question?

oshoval replied:
the one on the first commit

https://github.com/kubevirt/kubevirt/pull/14419/commits/b9ba1c71fd24ea9e4b79d7684e3a9da962418234#diff-72336e7282e99e07104813732305eeabe4fd5e9b9c13e7c689bbaf78c1afbd8aL50

orelmisan replied:
It could be removed and libvmi.WithAutoAttachPodInterface(false) will be added.
Is it ok with you?

orelmisan replied:
> default is auto attach, so i guess you just want it to be explicit ?

Yes.

oshoval replied:
i think we miss by adding it explicitly,
reader shoiuldn't care about it,
auto atttach is a feature of kubevirt itself, not a package that we use,  where it worth to be explicit

but wont insist

orelmisan commented on pkg/network/controllers/vm_test.go:
```diff
@@ -509,6 +509,70 @@ var _ = Describe("VM Network Controller", func() {
+					libvmistatus.WithInterfaceStatus(v1.VirtualMachineInstanceNetworkInterface{
+						Name:             netToDetachName,
+						PodInterfaceName: namescheme.GenerateHashedInterfaceName(netToDetachName),
+						InfoSource:       multusAndDomainInfoSource,
```
This is how the VMI is configured in a hotunplug scenario.

oshoval replied:
ye i know but for this test those vars arent required right ?
this is the point

orelmisan replied:
> ye i know but for this test those vars arent required right ?
this is the point

Define required.
I find it helps with understanding the tested scenario.

orelmisan replied:
Same answer as https://github.com/kubevirt/kubevirt/pull/14419#discussion_r2098236021.

oshoval replied:
not same case imho, but lets agree to disagree
hold can removed please, thanks

orelmisan commented on pkg/network/controllers/hotplug_test.go:
```diff
@@ -58,11 +58,5 @@ var _ = Describe("Network interface hot{un}plug", func() {
 			[]v1.Interface{{Name: "blue"}, {Name: "red"}},
 			[]v1.Network{{Name: "blue"}, {Name: "red"}},
 		),
-		Entry("should remain, given status iface and no associated iface in spec",
```
> I was hoping to `controllers.ClearDetachedInterfaces()` refactored because it is very hard to follow. While it's still here I'm having a hard time letting go of the direct tests that demonstrate its behavior. Since "education" does not justify keeping redundant code, I won't block on this, however I do recommend that we take this as a follow up.

The purpose of this PR is to ease refactoring the code in follow ups.

orelmisan commented:
Lets wait for an agreement.

/hold

orelmisan commented:
Removing the hold per @oshoval's agreement in https://github.com/kubevirt/kubevirt/pull/14419#discussion_r2098243723.
/hold cancel
